raiden_contracts/contracts/MonitoringService.sol

```diff
@@ -0,0 +1,321 @@
+pragma solidity ^0.4.23;
+
+import "./Token.sol";
+import "./Utils.sol";
+import "./lib/ECVerify.sol";
+import "./TokenNetwork.sol";
+
+contract MonitoringService is Utils {
+
+    // Token to be used for paying the rewards
+    Token public token; // Only allow RDN?
+
+    // The minimum allowed deposit for a MS
+    uint256 public minimum_deposit;
+
+    // keep track of registered Monitoring Services
+    // A Monitoring Services must deposit to register
+    mapping(address => bool) public registered_monitoring_services;
+
+    // channel_identifier => Struct
+    // Keep track of the rewards per channel
+    mapping(bytes32 => Reward) rewards;
+
+    // Keep track of balances
+    mapping(address => uint) public balances;
+
+    /*
+     *  Structs
+     */
+    struct Reward{
+        // The signature of the reward
+        bytes reward_proof_signature;
+
+        // Nonce of the most recently provided BP
+        uint256 nonce;
+
+        // Address of the Raiden Node that was monitored
+        // This is also the address that has the reward deducted from its deposit
+        address reward_sender_address;
+    }
+
+    /*
+     *  Events
+     */
+
+    //event MonitoringServiceNewDeposit(address sender, address receiver, uint amount);
+    event NewDeposit(address sender, address receiver, uint amount);
+    event RegisteredMonitoringService(address monitoring_service);
+    event NewBalanceProofReceived(
+        bytes reward_proof_signature,
+        uint256 nonce,
+        address ms_address,
+        address raiden_node_address
```

Comment:
This is called in `monitor`. I am not entirely sure that this is needed. Where do you use this currently?
If you really need it (e.g. restoring MS state), you can only emit `balance_proof_id` (`keccak256(channel_identifier, token_network_address)`), `ms_address`, because this should be the only thing needed to see what rewards you need to claim.

Reply:
this will be probably used when multiple MSs can update the channel - if the event is emited, it no longer makes sense to call `monitor`. emitting just the hash is ok 
(but I am not sure about costs - isn't `keccak256` more expensive than plain `emit`?)

Reply:
logs are `375` + `8` for each argument if not indexed. `keccak256` is 30 + `6` for each argument.